src/theory/quantifiers/ceg_t_instantiator.cpp

```diff
@@ -835,7 +842,174 @@ bool EprInstantiator::processEqualTerms( CegInstantiator * ci, SolvedForm& sf, N
+      return sv;
+    }else{
+      unsigned rmod = 0; // TODO : randomize?
+      for( unsigned i=0; i<lit.getNumChildren(); i++ ){
+        unsigned ii = ( i + rmod )%lit.getNumChildren();
+        Node litc = getPathToPv( lit[ii], pv, sv, path, visited );
+        if( !litc.isNull() ){
+          // path is outermost term index last
+          path.push_back( ii );
+          std::vector< Node > children;
+          if( lit.getMetaKind() == kind::metakind::PARAMETERIZED ){
+            children.push_back( lit.getOperator() );
+          }
+          for( unsigned j=0; j<lit.getNumChildren(); j++ ){
+            children.push_back( j==ii ? litc : lit[j] );
+          }
+          return NodeManager::currentNM()->mkNode( lit.getKind(), children );
+        }
+      }
+    }
+  }
+  return Node::null();
+}
+
+Node BvInverter::getPathToPv( Node lit, Node pv, Node sv, Node pvs, std::vector< unsigned >& path ) {
+  std::map< TNode, bool > visited;
+  Node slit = getPathToPv( lit, pv, sv, path, visited );
+  if( !slit.isNull() ){
+    // substitute pvs for the other occurrences of pv
+    TNode tpv = pv;
+    TNode tpvs = pvs;
+    slit = slit.substitute( tpv, tpvs );
+  }
+  return slit;
+}
+
+Node BvInverter::solve_bv_constraint( Node sv, Node sv_t, Node t, Kind rk, bool pol, std::vector< unsigned >& path,
+                                      BvInverterModelQuery * m, BvInverterStatus& status ) {
+  if( path.empty() ){
+    Assert( sv_t==sv );
+    // finalize based on the kind of constraint
+    //TODO
+    return t;
+  }else{
+    unsigned index = path.back();
+    Assert( index<sv_t.getNumChildren() );
+    path.pop_back();
+    Kind k = sv_t.getKind();
+    // inversions
+    if( k==BITVECTOR_PLUS ){
+      Node tt = NodeManager::currentNM()->mkNode( BITVECTOR_SUB, t, sv_t[1-index] );
+      return solve_bv_constraint( sv, sv_t[index], tt, rk, pol, path, m, status );
+    }else if( k==BITVECTOR_MULT ){
+      // TODO
+    }else if( k==BITVECTOR_NEG || k==BITVECTOR_NOT ){
+      Node tt = NodeManager::currentNM()->mkNode( k, t );
+      return solve_bv_constraint( sv, sv_t[index], tt, rk, pol, path, m, status );
+    }else if( k==BITVECTOR_AND || k==BITVECTOR_OR ){
+      // TODO
+    }else if( k==BITVECTOR_CONCAT ){
+      // TODO
+    }else if( k==BITVECTOR_SHL || k==BITVECTOR_LSHR ){
+      // TODO
+    }else if( k==BITVECTOR_ASHR ){
+      // TODO
+    }else{
+      Trace("bv-invert") << "bv-invert : Unknown kind for bit-vector term " << sv_t << std::endl;
+    }
+    return Node::null();
+  }
+}
+
+Node BvInverter::solve_bv_lit( Node sv, Node lit, bool pol, std::vector< unsigned >& path,
+                               BvInverterModelQuery * m, BvInverterStatus& status ){
+  Assert( !path.empty() );
+  unsigned index = path.back();
+  Assert( index<lit.getNumChildren() );
+  path.pop_back();
+  Kind k = lit.getKind();
+  if( k==NOT ){
+    Assert( index==0 );
+    return solve_bv_lit( sv, lit[index], !pol, path, m, status );
```

Comment:
Do this iteratively, to avoid issues with deeply nested `NOT` chains? Or are they guaranteed to be preprocessed away?